File: trees/binarysearchtree/binarysearchtree.py

class node :

	def __init__(self , data = None) :

		self.data  = data
		self.left  = None
		self.right = None


class binarysearchtree :

	def __init__(self) :

		self.root = node()


	def _append(self , value) :
	
		newnode = node(value)
		if self.root.data == None :
			self.root = newnode
			return

		temp = self.root
		parent = None

		while temp != None :

			parent = temp
			if value <= temp.data :
				temp = temp.left
			else :
				temp = temp.right

		if value <= parent.data :
			parent.left = newnode
		else :
			parent.right = newnode

	def _max(self) :

		temp = self.root
		while temp.right != None :
			temp = temp.right

		return temp.data

	def _min(self) :

		temp = self.root
		while temp.left != None :
			temp = temp.left

		return temp.data

	def _breathfirst(self) :

		if self.root == None :
			return

		q = []
		array = []
		q.append(self.root)

		while len(q) != 0 :
			current = q[0]
			array.append(current.data)

			q.pop(0)

			if current.left != None :
				q.append(current.left)
			if current.right != None :
				q.append(current.right)

		return array


	def _search(self , value) :

		if self.root == None :
			return
		temp = self.root
		while temp != None :
			if temp.data == value :
				return True
			else :
				if value <= temp.data :
					temp = temp.left
				else :
					temp = temp.right

		return False

	def _hight(self) :

		def _highthelper(temp) :
			if temp == None:
				return -1
			else :
				leftsubtree  = _highthelper(temp.left)
				rightsubtree = _highthelper(temp.right)

			return 1 + max(leftsubtree, rightsubtree)

		if self.root == None :
			return -1
		else :
			return _highthelper(self.root)

	def _postorder(self) :

		s1 = []
		s2 = []
		array = []

		temp = self.root

		s1.append(temp)

		while len(s1) != 0 :

			temp = s1[-1]
			s1.pop()
			s2.append(temp)

			if temp.left :
				s1.append(temp.left)
			if temp.right :
				s1.append(temp.right)

		while len(s2) != 0 :

			temp = s2[-1]
			s2.pop()
			array.append(temp.data)

		return array

	def _inorder(self) :

		if self.root == None :
			return

		temp = self.root
		stack = []
		array = []

		while temp != None or not len(stack) == 0 :

			while temp != None :

				stack.append(temp)
				temp = temp.left

			temp = stack[-1]
			stack.pop()
			array.append(temp.data)
			temp = temp.right

		return array

	def _preorder(self) :

		if self.root == None :
			return

		stack = []
		array = []
		temp = self.root

		while temp != None or not len(stack) == 0 :
			while temp != None :
				array.append(temp.data)
				if temp.right :
					stack.append(temp.right)
				temp = temp.left
			if not len(stack) == 0 :
				temp = stack[-1]
				stack.pop()

		return array 

bst = binarysearchtree()

bst._append(15)
bst._append(6)
bst._append(3)
bst._append(9)
bst._append(8)
bst._append(20)
bst._append(25)
bst._append(30)
bst._append(40)



print(f"Max    : {bst._max()}")
print(f"Min    : {bst._min()}")
print(f"Hight  : {bst._hight()}")
print(f"Search : {bst._search(6)}")

print(bst._breathfirst())
print(bst._postorder())
print(bst._inorder())
print(bst._preorder())

File: trees/binarysearchtree/binarysearchtree.cpp

#include <bits/stdc++.h>

using namespace std ;

template <typename T>

class bst
{
	private :

		struct node
		{
			T d ;
			node *left ;
			node *right ;
		} ;

		node *root = nullptr ;
		int _highthelper(node *temp)
		{
			if(temp == nullptr) // empty sub tree
				return -1 ;

			int leftsubtree = _highthelper(temp->left) ;
			int rightsubtree = _highthelper(temp->right) ;

			return 1 + max(leftsubtree , rightsubtree) ;
		}

		void _postorderhelper(node *temp)
		{
			if(temp == nullptr)
				return ;
			_postorderhelper(temp->left) ;
			_postorderhelper(temp->right) ;
			cout << temp->d << " " ;
		}

	public :

		void _append(T v) ;
		int _max() ;
		int _min() ;
		int _hight() ;
		void _breathfirst() ;
		void _postorder() ;
		void _inorder() ;
		void _preorder() ;
		bool _search(T v) ;
} ;

template <typename T>

void bst<T>::_append(T v) // leaf node (last node on a tree)
{
	node *newnode = new node ;
	newnode->d = v ;
	newnode->left = nullptr ;
	newnode->right = nullptr ;

	if(root == nullptr)
	{
		root = newnode ;
		return ;
	}
	else
	{
		node *temp = root ;
		node *parent = nullptr ;

		while(temp != nullptr)
		{
			parent = temp ;
			if(v <= temp->d)
				temp = temp->left ;
			else
				temp = temp->right ;
		}

		if(v <= parent->d)
			parent->left = newnode ;
		else
			parent->right = newnode ;
	}
}

template <typename T>

int bst<T>::_max()
{
	node *temp = root ;

	while(temp->right != nullptr)
		temp = temp->right ;

	return temp->d ;
}

template <typename T>

int bst<T>::_min()
{
	node *temp = root ;

	while(temp->left != nullptr)
		temp = temp->left ;

	return temp->d ;
}

template <typename T>

int bst<T>::_hight()
{
	if(root == nullptr)
		return -1 ;
	else
		return _highthelper(root) ;
}

template <typename T>

void bst<T>::_breathfirst()
{
	if(root == nullptr)
		return ;
	else
	{
		queue<node *> q ;
		q.push(root) ;

		cout << "[ " ;
		while(! q.empty())
		{
			node *current = q.front() ;

			cout << current->d << " " ;

			q.pop() ;

			if(current->left != nullptr)
				q.push(current->left) ;
			if(current->right != nullptr)
				q.push(current->right) ;
		}
		cout << "]" << endl ;
	}
}

template <typename T>

void bst<T>::_postorder()

{
	if (root != nullptr)
		cout << "[ " ;
		_postorderhelper(root) ;
		cout << "]" << endl ;
}

template <typename T>

void bst<T>::_inorder()
{
	if(root == nullptr)
		return ;

	node *temp = root ;
	stack<node*> s ;

	cout << "[ " ;
	while(temp != nullptr || ! s.empty())
	{
		while(temp != nullptr)
		{
			s.push(temp) ;
			temp = temp->left ;
		}

		temp = s.top() ;
		s.pop() ;
		cout << temp->d << " " ;
		temp = temp->right ;
	}
	cout << "]" << endl ;
}

template <typename T>

void bst<T>::_preorder()
{
	if(root == nullptr)
		return ;

	stack<node*> s ;
	node *temp = root ;

	s.push(temp) ;

	cout << "[ " ;
	while(temp != nullptr and s.empty() == false)
	{
		while(temp != nullptr)
		{
			cout << temp->d << " " ;
			if(temp->right)
				s.push(temp->right) ;
			temp = temp->left ;
		}

		if(s.empty() == false)
		{
			temp = s.top() ;
			s.pop() ;
		}
	}
	cout << "]" << endl ;
}

template <typename T>

bool bst<T>::_search(T v)
{
	if(root == nullptr)
		return false ;
	else
	{
		node *temp = root ;
		while(temp != nullptr)
		{
			if(temp->d == v)
				return true ;
			else
			{
				if(temp->d >= v)
					temp = temp->left ;
				else
					temp = temp->right ;
			}
		}
	}
	return false ;
}

int main(void)
{
	bst<int> t ;

	t._append(15) ;
	t._append(6) ;
	t._append(3) ;
	t._append(9) ;
	t._append(8) ;
	t._append(20) ;
	t._append(25) ;
	cout << "min    : " << t._min() << endl ;
	cout << "max    : " << t._max() << endl ;
	cout << "hight  : " << t._hight() << endl ;
	cout << "search : " << t._search(25) << endl ;
	t._breathfirst() ;
	t._postorder() ;
	t._inorder() ;
	t._preorder() ;
	return 0 ;
}